pkg/controller/velero/velero.go

```diff
@@ -0,0 +1,233 @@
+	if err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: "default"}, foundVsl); err != nil {
+		if errors.IsNotFound(err) {
+			// Didn't find VolumeSnapshotLocation
+			reqLogger.Info("Creating VolumeSnapshotLocation")
+			if err := controllerutil.SetControllerReference(instance, vsl, r.scheme); err != nil {
+				return reconcile.Result{}, err
+			}
+			if err = r.client.Create(context.TODO(), vsl); err != nil {
+				return reconcile.Result{}, err
+			}
+		} else {
+			return reconcile.Result{}, err
+		}
+	} else {
+		// VolumeSnapshotLocation exists, check if it's updated.
+		if !reflect.DeepEqual(foundVsl.Spec, vsl.Spec) {
+			// Specs aren't equal, update and fix.
+			reqLogger.Info("Updating VolumeSnapshotLocation")
+			foundVsl.Spec = *vsl.Spec.DeepCopy()
+			if err = r.client.Update(context.TODO(), foundVsl); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+	}
+
+	// Install CredentialsRequest
+	foundCr := &minterv1.CredentialsRequest{}
+	cr := credentialsRequest(namespace, credentialsRequestName, instance.Status.S3Bucket.Name)
+	if err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: credentialsRequestName}, foundCr); err != nil {
+		if errors.IsNotFound(err) {
+			// Didn't find CredentialsRequest
+			reqLogger.Info("Creating CredentialsRequest")
+			if err := controllerutil.SetControllerReference(instance, cr, r.scheme); err != nil {
+				return reconcile.Result{}, err
+			}
+			if err = r.client.Create(context.TODO(), cr); err != nil {
+				return reconcile.Result{}, err
+			}
+		} else {
+			return reconcile.Result{}, err
+		}
+	} else {
+		// CredentialsRequest exists, check if it's updated.
+		if !reflect.DeepEqual(foundCr.Spec, cr.Spec) {
+			// Specs aren't equal, update and fix.
+			reqLogger.Info("Updating CredentialsRequest")
+			foundCr.Spec = *cr.Spec.DeepCopy()
+			if err = r.client.Update(context.TODO(), foundCr); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+	}
+
+	// Install Deployment
+	foundDeployment := &appsv1beta1.Deployment{}
+	deployment := veleroInstall.Deployment(namespace, veleroInstall.WithoutCredentialsVolume(),
+		veleroInstall.WithEnvFromSecretKey(strings.ToUpper(awsCredsSecretIDKey), credentialsRequestName, awsCredsSecretIDKey),
+		veleroInstall.WithEnvFromSecretKey(strings.ToUpper(awsCredsSecretAccessKey), credentialsRequestName, awsCredsSecretAccessKey),
+		veleroInstall.WithImage("gcr.io/heptio-images/velero:v1.0.0"),
+	)
+	if err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: "velero"}, foundDeployment); err != nil {
+		if errors.IsNotFound(err) {
+			// Didn't find Deployment
+			reqLogger.Info("Creating Deployment")
+			if err := controllerutil.SetControllerReference(instance, deployment, r.scheme); err != nil {
+				return reconcile.Result{}, err
+			}
+			if err = r.client.Create(context.TODO(), deployment); err != nil {
```

Comment:
Before creating the `Deployment`, thoughts on `nodeAffinity` for v1.0.0 to require amd64 arch `Node`s? I don't know to what extent we will be running on arm64 for 4.x in EC2 for customers, but velero (this version anyways) does not support anything but amd64/linux:

```lisa@laptop ~ $ docker run --rm -ti mplatform/mquery gcr.io/heptio-images/velero:v1.0.0
Image: gcr.io/heptio-images/velero:v1.0.0
 * Manifest List: No
 * Supports: amd64/linux
```

Reply:
Not a bad idea. Let me see what I can do.